pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,281 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgtype"
+
+	"github.com/prometheus/client_golang/prometheus"
+	io_prometheus_client "github.com/prometheus/client_model/go"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgmodel/common/schema"
+	"github.com/timescale/promscale/pkg/pgmodel/model/pgutf8str"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+type engine struct {
+	uuid [16]byte
+	conn pgxconn.PgxConn
+
+	stop chan struct{}
+
+	metrics sync.Map
+}
+
+type Telemetry interface {
+	Start()
+	// RegisterMetric registers a Prometheus Gauge or Counter metric for telemetry visitor.
+	// It must be called after creating the telemetry engine.
+	RegisterMetric(columnName string, gaugeOrCounterMetric prometheus.Metric) error
+	Stop()
+	// Sync syncs the registered metric values with the information table. This is used only
+	// in end-to-end tests.
+	Sync() error
+}
+
+func NewTelemetryEngine(conn pgxconn.PgxConn, uuid [16]byte) (Telemetry, error) {
+	isTelemetryOff, err := isTelemetryOff(conn)
+	if err != nil {
+		log.Debug("msg", "unable to get TimescaleDB telemetry configuration. Maybe TimescaleDB is not installed", "err", err.Error())
+	}
+	if !isTelemetryOff {
+		// Warn the users about telemetry collection only if telemetry collection is enabled.
+		log.Warn("msg", "Promscale collects anonymous usage telemetry data to help the Promscale team better understand and assist users. "+
+			"This can be disabled via the process described at https://docs.timescale.com/timescaledb/latest/how-to-guides/configuration/telemetry/#disabling-telemetry")
+	}
+	engine := &engine{
```

Comment:
Nit: this can really hurt you, shadowing struct names with variable names. I usually take care not to do that.

Just to avoid any confusion when reading the code.